### PR TITLE
Testtrain

### PR DESCRIPTION
**Pull request by nirveshk**
Submitting a new pull request due to bad initial pull request. 

Supersedes #27 and implement #11

### CODE REVIEW

File: hs628_tools/Test.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+    #set path to read the csv file in the directory
+    def test_train():
+
+        pathName = os.getcwd()
+
+        dataset = []
+        fileNames = os.listdir(pathName)
+
+        # read file with extension .csv
```
Comment (omidkj) [inline hs628_tools/Test.py]: Are you trying to save all the files in that directory with CSV format? 

```python
for fileName in fileNames:
   if fileName.endswith(".csv"):
      dataset.append(fileName)
```

File: hs628_tools/Test.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+        for fileNames in fileNames:
+            if fileNames.endswith(".csv"):
+                dataset.append(fileNames)
+
+        for k in dataset:
```
Comment (omidkj) [inline hs628_tools/Test.py]: If you want to clean all the CSV files in that directory, you'll need to save each dataset in a list and then proceed with the cleaning part. The current code is looping through the CSV files and read the file and save it into 'data' variable and each time the 'data' will be entirely replaced with a new CSV file content. Therefore, you are only saving the very last CSV file in your folder.

Comment (nirveshk) [reply hs628_tools/Test.py]: That is correct. This part of the code is taking in csv file and reading it in. Looping through each data point is the next step. I hope you saw that as well. Please let me know if you didn't.

File: hs628_tools/Test.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    row_len = sum(1 for line in test_train())
+    print row_len
+
+
+    def split_ratio(row_len):
+        """
+
+        :return: data split ratio based on the number of data points
+        """
```
Comment (NikitaThomas) [inline hs628_tools/Test.py]: It would be helpful to add in the docstrings what data type the output is that you are returning. I believe your code right now is just print statements so not returning any data type. I think it would be useful to instead return the ratio into a list especially for my visualization module because I want to access the test ratio calculated from this module in order to split the dataset and run my code.

File: hs628_tools/Test.py [hunk 4]
```diff
@@ -38,26 +38,49 @@ def test_train():
...
+        if row_len <= 500:
```
Comment (haleyhowe) [inline hs628_tools/Test.py]: How are these metrics being determined? Do you have reasoning as to why you chose these upper and lower quartiles?

Comment (nirveshk) [reply hs628_tools/Test.py]: Hi Haley,
The split ratio were assigned based on the length of the data. We all know there is no hard and fast rule to what the ratio should be, but I came up with, larger the data larger should be the training date set. I believe, this makes sense because we can leverage the voluminous  data to train more for better chances of model accuracy and in contrary, train comparatively smaller portion of data in smaller data  set so to leave enough test cases to overrule any over fittings. Again, this may not be the best answer to why I decided to choose split ratios in such manner but I hope I was able to communicate my intention. Thanks!

Nirvesh

File: hs628_tools/Test.py (hunk 3 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: Hi Nikita,
Thank you for your suggestion and insight. I have updated my new codes with docstrings. I will make sure to word the docstring in accordance with what the codes actually do. That makes sense. I have completely changed my codes to have it print new train and test dataframes for the users. Please take a look and let me know if I can do anything to polish it. Thanks again!

Nirvesh

**Review by kamehta2 — COMMENTED** [review]
Code written for Test and Train data ratios can benefit by being a separate function. And then another function to build the dataframes for test and train data which also include both Nan and non-Nan data. Currently, you have only one main function that does the overall implementation. By breaking it further, it may help the user to choose the functionality they are interested in.

File: hs628_tools/Test.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+    defines method to accomplish this task.
+
+    returns a split ratio.
+
+    Identifies Missing values as Nan for any missing value, return train and test dataframe with uniform division
+    of NAN between train and test data per the data split ratio.
+
+    """
+    def __init__(self, data):
+        self.data = data
```
Comment (rohitchadaram) [inline hs628_tools/Test.py]: There is no reason to have self.data = data as you cannot use the self.data datatype further as it is not a default datatype in python. Remember the way we used to work with a Vector/Matrix data type in 608. So usually either you create a work around in your class to adhere to the new datatype for data or just add any other numerical/string values parameters into help.

File: hs628_tools/Test.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+            train = 75
+            test = 25
+        else:
+            train = 80
+            test = 20
+        print 'train:',train, 'test:',test
+
+
+        # drops all NAN values from data
+        non_nandata = data[~np.isnan(data).any(axis=1)]
```
Comment (rohitchadaram) [inline hs628_tools/Test.py]: It is generally not advisable to straight drop all NaN's primarily because you are dropping the entire row, with other column data which might not have a NaN. It should be evaluated based on the percentage of missing values for the column you can decide to impute the data rather than dropping. But anyways this feature doesn't really focus on that aspect so it's okay. But when dealing with small datasets which generally the case with health care data always think of a way to impute without creating bias rather than making them NaN's and dropping the rows.

File: hs628_tools/Test.py [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+
+
+    def test_train():
+        """
+        Reads in csv file using np.genfromtxt, counts number of NANs in the dataset
+
+        """
+
+        data = np.genfromtxt(csv_datafile, delimiter=',', skip_header=1)
+        np.isnan(data).sum() #counts the number of "NANs" in dat
```
Comment (hhan14) [inline hs628_tools/Test.py]: Hi Nirvesh, I just wanted to ask you whether you have a NameError for csv_datafile on line 32 as csv_datafile is not defined here. 
Also, this can be relatively minor point, but would like to suggest that you`d better build up your code upon the "PEP 8" guideline as your keep leaving lot of blank lines between codes. When you put your code on Pycharm, then you`ll see the difference as Pycharm automatically edits and arrange your code based on the PEP 8. I`ll share the link here which prof. Eric shared during the class for your reference. (https://www.python.org/dev/peps/pep-0008/)

Comment (RoxanneXin) [inline hs628_tools/Test.py]: Hi @nirveshk, I'd like your idea of split data. And I may use your idea to read my data instead my initial way. 
Just, make sure, so for the csv_datafile here, will it be the parameter of your function?

Comment (nirveshk) [reply hs628_tools/Test.py]: Hi Helen,
Thanks for your suggestion regarding "PEP 8". I have reviewed and will follow the instruction.
Regarding the csv_datafile, I tested the code using a dataset but chose to have csv_datafile instead(This is way of asking user to pass in a csv file). Thanks!

Nirvesh

Comment (nirveshk) [reply hs628_tools/Test.py]: Hi @RoxanneXin,

Thank you for considering my codes to be useful. Regarding the csv_datafile, it is actually meant to ask user to pass in a csv file they would like to work on and yes it is one of the parameters for the function. Thanks!

Nirvesh

File: hs628_tools/Test.py (hunk 5 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: Hi Rohit,

Thanks for your input, I should have had been careful before pushing the code. Thanks for pointing out this to me. You are right, I will go ahead and remove this portion. It was pushed in error. Thanks!

File: hs628_tools/Test.py (hunk 6 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: That makes sense. I was very focused in accomplishing the test and train split ratio and may have overlooked the sensitiveness of healthcare data. I will work on getting this included in my codes. Thanks.

-Nirvesh

File: hs628_tools/Test.py [hunk 8]
```diff
@@ -0,0 +1,133 @@
+
+import numpy as np
+from sklearn.model_selection import train_test_split
+
+
+
+
+
+
+class data_split:
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: Class names should be CapWords/CamelCase style

File: hs628_tools/Test.py (hunk 5 above)
Comment (EHWUSF) [reply hs628_tools/Test.py]: I'm not really understanding @rohitchadaram 's comment about "cannot use the self.data datatype further as it is not a default datatype in python. "  Currently there is nothing really explicitly specified about the datatype of ```data```, but if you specify in the docstring that a numpy ndarray needs to be provided, or a matrix-like object having methods X and Y, then you can reasonably rely on a user providing an object meeting those constraints (best to test before absolutely relying in order to avoid exceptions, but your main functionality could be dependent on a compliant object.

In this particular case, though, you are not currently trying to use the data later, so there doesn't seem to be a reason to have the user provide it here.  And from your response it sounds like it is your plan to remove it here, so just take this as a reminder...

File: hs628_tools/Test.py (hunk 7 above)
Comment (EHWUSF) [reply hs628_tools/Test.py]: @nirveshk as the above comments point out, this reference to ```csv_datafile``` is not yet a parameter and currently causes a name error if anyone tries to run this code.  In your response you speak in the present tense as if you have already made this a parameter, but that is not currently the case (or at least you have not pushed those changes here).  That would have been a good (and easy!) bit of cleanup to take care of right away.

But I'll also repeat the advice I've given in many of the other PRs so far-- there can be multiple reasons why the user might not have a data file with their data-- it's best to simply require the user to pass a data object (meeting whatever constraints you need to require), and let the user deal with the details of whether they are reading from some local file or getting their data some other way and/or processing it before sending it to your code.

Of course, for your test code you are the user, and need to deal with those details for your test data, but that should be a separate routine, not left woven into this method.

File: hs628_tools/Test.py [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+
+
+        #set calculation standard for specific datalengths, this is based on the general rule of thumb
+        #ratio increases as the length of the data increases, so to ensure we have  enough data to train and test
+        #for the smaller datasets
+        
+        if len(data) <= 500:
+            train = 60
+            test = 40
+        elif len(data) > 500 and row_len <= 10000:
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: ```row_len``` has not yet been defined.

File: hs628_tools/Test.py (hunk 6 above)
Comment (EHWUSF) [reply hs628_tools/Test.py]: It seems to me that the real issue here is that the comment in the code is so terse that, just looking at this point, it gives the wrong impression that the samples with NaN's are actually being thrown away, whereas really they are simply being dealt with separately.

@nirveshk your response to Rohit's comment makes it look like you yourself forgot what you were doing here, I think you must have rushed your response a bit too much.  But that just emphasizes that the comment in the code should be better.

File: hs628_tools/Test.py [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        # drops all NAN values from data
+        non_nandata = data[~np.isnan(data).any(axis=1)]
+
+        #print non_nandata.shape
+
+        #print non_nandata.shape
+
+
+
+        ### nan dataframe
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: You are not really making a dataframe here, and the array you make is not 100% NaN's, the comment should be more clear.

File: hs628_tools/Test.py [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+
+
+
+        #set calculation standard for specific datalengths, this is based on the general rule of thumb
+        #ratio increases as the length of the data increases, so to ensure we have  enough data to train and test
+        #for the smaller datasets
+        
+        if len(data) <= 500:
+            train = 60
+            test = 40
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: A somewhat minor point (but seems just important enough for me to mention) ```train``` and ```test``` are a bit generic names, they could be referring to the training and test data.  I think it would make the code slightly more clear if they were named ```train_percent``` and ```test_percent``` or perhaps let them be floating numbers and name them ```train_portion``` and ```test_portion```  (if you even keep the test variable, see below)

File: hs628_tools/Test.py [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+        elif len(data) > 500 and row_len <= 10000:
+            train = 70
+            test = 30
+        elif len(data) > 10000 and row_len <=100000:
+            train = 75
+            test = 25
+        else:
+            train = 80
+            test = 20
+        print 'train:',train, 'test:',test
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: This seems like a bit of testing code that got left behind.  If you are intending it to stay then the output should be more clear (and should be mentioned in the docstring).

File: hs628_tools/Test.py [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+        nan_data = data[np.isnan(data).any(axis = 1)]
+        #print nan_data
+
+        #####split  Nan data and non_nandata using the split ratio
+        #####determined above
+
+
+
+
+        int_train = (train/float(100))
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: This is named as an integer, but since you cast the 100 to float the result is not actually integer.  And you shouldn't even cast an explicit int to float when you can instead simply add a decimal: ```100.0```

File: hs628_tools/Test.py [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+            test = 40
+        elif len(data) > 500 and row_len <= 10000:
+            train = 70
+            test = 30
+        elif len(data) > 10000 and row_len <=100000:
+            train = 75
+            test = 25
+        else:
+            train = 80
+            test = 20
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: Other than your print statement, you don't wind up actually using these ```test``` values, you instead just determine the corresponding number from the train value (which is the right approach from a reliability standpoint, so just don't even declare these ```test``` values)

File: hs628_tools/Test.py [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+
+
+
+        int_train = (train/float(100))
+        #print int_train
+
+        ##split Nan data to train and test sets
+
+        train_nan_data = nan_data[:int(train/float(100)*len(nan_data)), :]
+        test_nan_data = nan_data[int(train/float(100)*len(nan_data)):, :]
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: The cast to float could be avoided by instead simply reordering the calculation: ```int(train*len(nan_data)/100)```.  I leave the cast to int in order to improve compatibility between python 2.7 and 3.x, though it might be slightly better to instead use explicit integer division: ```train*len(nan_data)//100```

File: hs628_tools/Test.py [hunk 16]
```diff
@@ -0,0 +1,133 @@
...
+        #print train_nan_data.shape
+        #print test_nan_data.shape
+
+
+
+        ##split non Nan data to train and test sets
+        train_non_nan = non_nandata[:int(train/float(100)*len(non_nandata)), :]
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: same comments as with the nan_data

File: hs628_tools/Test.py [hunk 17]
```diff
@@ -0,0 +1,133 @@
...
+
+        #### Add test_non_nan and test_nan_data
+
+        final_test = np.concatenate([test_non_nan, test_nan_data])
+        #print final_test.shape
+
+
+
+
+        ##This module prints out the train and test data as per the preset ratio
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: As I think someone else noted already, but is important enough for me to repeat since you haven't yet fixed this-- printing out a potentially huge amount of data is not especially helpful for the user; what they really need is to get returned a tuple with the respective datasets.

Also, the text of this comment (with suitable modification) is really much more important as part of the docstring than it is here.

File: hs628_tools/Test.py [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+        ## determined based on the length of the data and includes missing data
+        ## uniformly between train and test data
+        print "This is the final train data:"\
+                ,final_train
+        print "This is the final test data:" \
+            , final_test
+
+    test_train()
```
Comment (EHWUSF) [inline hs628_tools/Test.py]: As we've discussed in class, your module should not have any side effects (even simply printing out a large block of data) get triggered by merely importing it-- this call (which is really a bit of test code) should get wrapped into a ```if __name__ == "__main__:" block (and this is then the place to read in your test data and provide it to your routine).   And since you've defined a Class here, your test code should make an instance of that class.

File: hs628_tools/Test.py (hunk 8 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: Class name has been changed as per the suggestion. Thank you!

File: hs628_tools/Test.py (hunk 5 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: @EHWUSF That is correct, since I do not intend to use data later, I decided to omit this part from my package.

File: hs628_tools/Test.py (hunk 7 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: @EHWUSF I have hard coded my datasets while reading it with np.genfromtxt. I have also added a url link to where that data is available(in my github account under HS628 project1 repo). This way user can either download the data from the link provided or add their own data to use the python package I created. I read this suggestion in one of your response to one of our classmate's pull request and I decided to go with it. Thanks!

File: hs628_tools/Test.py (hunk 11 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: @EHWUSF  I have applied your recommendation and updated the codes. Thank you!

File: hs628_tools/Test.py (hunk 9 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: @EHWUSF row_len has been defined. Thanks!

File: hs628_tools/Test.py (hunk 14 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: @EHWUSF Understood and complied with your recommendation. Thanks!

File: hs628_tools/Test.py (hunk 10 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: Modified comment per the recommendation to make it more clear to the users.

File: hs628_tools/Test.py (hunk 12 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: @EHWUSF necessary changes per the recommendation has been made. Added "Suggests train and test ratio and prints out the train_percentage for user's convenience" to the docstring.

File: hs628_tools/Test.py (hunk 6 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: @EHWUSF I agree with your opinion, I may have rushed in replying @rohitchadaram. I was under the impression that the Nan data were actually dropped. However, later I realized this is not the case in my package. I am temporarily dropping rows with NAN values so that I can accumulate all NAN rows in an array and later use them to uniformly distribute(based on the train test ratio) across train and test set without losing a single piece of information.

File: hs628_tools/Test.py (hunk 13 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: @EHWUSF That makes sense. Adding simply a decimal after the value indeed avoids the need to cast the value to float. Thanks!

File: hs628_tools/Test.py (hunk 15 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: Made the necessary changes per the recommendation. Thanks!

File: hs628_tools/Test.py (hunk 16 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: Made the necessary changes per the recommendation. Thanks!

File: hs628_tools/Test.py (hunk 17 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: Made modification to have the function return the train and test datasets as tuple. Added docstring to reflect that the function returns train and test datasets as tuple.

File: hs628_tools/Test.py (hunk 18 above)
Comment (nirveshk) [reply hs628_tools/Test.py]: Thank you for all your constructive response. I have wrapped my codes in "if __name__ == "__main__"per your suggestion and read the test data here instead in the user defined function. 
Thanks,
Nirvesh
